oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,370 @@
+- A decision on whether we use the MFE config API, env.config.js, both, or something else to supply the module federation configuration, whether it's one big combined document or whether each MFE has its own.
+- How we sandbox and put error boundaries around dynamically loaded modules.
+- Whether we create a central global state store, and whether that's redux.
+- Whether we need an eventing system for inter-MFE communication.
+- How we manage breaking dependency changes across MFEs.
+
+Rejected Alternatives
+*********************
+
+Piral
+=====
+
+A prior iteration of this OEP and discovery effort (FC-0007) came to the conclusion that we should adopt Piral, a comprehensive micro-frontend web framework, to address our concerns with the Open edX micro-frontend architecture.
+
+After further investigation and review of our stated pains, observed deficiencies, hopes, and vision for Open edx micro-frontends, we chose to adjust course away from Piral. Piral solves runtime module loading and shared dependencies in a similar way to Webpack module federation - and can in fact use it internally - but does so in a more proprietary, opinionated, and opaque way, adding additional layers/wrappers around it. While Webpack is one of many bundlers available, it's also the defacto standard against which others are judged, and has wide industry adoption. Webpack module federation is a standard approach for composing frontends.
+
+Piral is an impressive piece of software, built primarily by one individual, trying to solve a much broader problem than we have. Because of this, it brings along with it a great deal of complexity that we don't need and already have solutions for. Piral aims to be a complete toolkit for building web applications, including authentication, plugins, its own global state mechanism, extensions that provide ready-made UI components, etc.
+
+We need a mechanism to provide shared dependencies and composable frontends that can fit in with our existing ecosystem. Adopting Piral would likely involve significant refactoring of existing MFEs to fit into its framework and to turn them into "pilets", which locks us in to the Piral way of doing things.
+
+It feels like our needs more closely align with the narrower scope of module federation, and that it's a more right-sized solution to our architectural problems.
+
+Combining MFEs
+==============
+
+Folding our micro-frontends together into a few larger frontends (LMS and Studio, for instance) solves our need for shared dependencies in a different way - it just shares all the code so there's one set of dependencies for all of it. We could continue to rely on frontend-plugin-framework for cross-domain plugins, but "plugins" within the larger domain become a simple import from another part of the application.
+
+This approach was abandoned because we still believe that MFE independence is a core need for our platform, and we can't go back to a few monolithic frontends.
```

Comment:
This also wouldn't get us runtime module replacement without iframes right?

Reply:
I don't think this approach does anything useful for that, yeah... you'd be stuck with using FPF, getting a slot added to the platform for the thing you want to replace, and then doing it with an iframe out to your custom MFE.